### DISCUSSION TITLE
Dynamically generate robot.txt and sitemap.xml based on client hostname

### DISCUSSION THREAD

**Discussion by charinduedirisuriya** [vercel/next.js#18697] [Help]
I have a very unique requirement;
We need to dynamically generates two files, with two different content-types (xml and txt) to serve our robots.txt and sitemap.xml.

The catch here is that both these files MUST be served from the root directory and should be served dynamically because we need to inspect the hostname of the application and render the two files based on the hostname.

Example different hosts can have different sitemaps; therefore it is needed to fetch the sitemaps from the server and render them accordingly to the client.

I approached this using the `getInitialProps` method but unfortunately the hostname is not accessing as this code is built on the server side and not on the client.

```
class Sitemap extends React.Component {

    static async getInitialProps({ res }) {

        if(res) {

            const hostName= res.headers.host;
            console.log(`request_source ${hostName}`);

           // do some other works and get data and render this accordingly 

            if (isSubDomain(hostName)) {

                 // just for testing I have hardcoded this
                const textFile = () => {
                    return `User-agent: *
                        Disallow: /`;
                };
                res.setHeader("Content-Type", "text/plain;charset=UTF-8");
                res.write(textFile());
                res.end();

            } else {

                const textFile = () => {
                    return `User-agent: *
                        Disallow:`;
                };
                res.setHeader("Content-Type", "text/plain;charset=UTF-8");
                res.write(textFile());
                res.end();

            }
        }
    }
}

export default Sitemap;
```

**Comment by wyantb** [comment]
The request, not the response, has host info available. Try:

```
static async getInitialProps({ req, res }) {
  console.log(req.headers.host);
  ...
}
```

Otherwise, the skeleton of your solution was quite useful (I searched "nextjs generated robots.txt"), I didn't think think to use the `res.*` methods from within getInitialProps like this. Cheers!
